File: bindings/ruby/lib/pls4all.rb
# SPDX-License-Identifier: CeCILL-2.1
#
# Ruby binding around libp4a's `p4a_pls_fit_simple` C ABI helper.
# Uses Fiddle from the Ruby standard library — no extra gems required.
# Parity-gated against the shared cross-binding fixture
# `bindings/js/test/parity_fixture.json` at machine epsilon.

require "fiddle"
require "fiddle/import"

module Pls4all
  # Internal: Fiddle::Importer wired to libp4a. Public callers should
  # use the module-level `version`, `abi_version` and `pls_fit` methods.
  module Lib
    extend Fiddle::Importer

    candidates = []
    candidates << ENV["PLS4ALL_LIB"] if ENV["PLS4ALL_LIB"] && !ENV["PLS4ALL_LIB"].empty?
    candidates += [
      "libp4a.so",
      "libp4a.dylib",
      "p4a.dll",
    ]
    # Walk up from this file to find a CMake dev-release build.
    here = File.expand_path(File.dirname(__FILE__))
    until here == "/" || here.empty?
      %w[libp4a.so libp4a.dylib].each do |name|
        path = File.join(here, "build/dev-release/cpp/src", name)
        candidates << path if File.exist?(path)
      end
      parent = File.dirname(here)
      break if parent == here
      here = parent
    end

    loaded = false
    candidates.each do |path|
      next if path.nil? || path.empty?
      begin
        dlload path
        loaded = true
        break
      rescue Fiddle::DLError
        next
      end
    end
    raise LoadError, "could not load libp4a (tried: #{candidates.inspect})" unless loaded

    extern "const char* p4a_get_version_string(void)"
    extern "unsigned int p4a_get_abi_version_major(void)"
    extern "unsigned int p4a_get_abi_version_minor(void)"
    extern "unsigned int p4a_get_abi_version_patch(void)"
    extern "int p4a_pls_fit_simple(const void*, const void*, int, int, int, int, " \
           "void*, void*, void*, void*)"
  end
  private_constant :Lib

  # Returns the libp4a runtime version, e.g. "0.91.0+abi.1.13.0".
  def self.version
    ptr = Lib.p4a_get_version_string
    return "" if ptr.null?
    ptr.to_s
  end

  # Returns [major, minor, patch] from the libp4a ABI.
  def self.abi_version
    [Lib.p4a_get_abi_version_major,
     Lib.p4a_get_abi_version_minor,
     Lib.p4a_get_abi_version_patch]
  end

  # Bundle of arrays returned by {Pls4all.pls_fit}.
  FitResult = Struct.new(
    :n, :p, :q, :n_components,
    :coefficients,  # length p*q (row-major)
    :x_mean,        # length p
    :y_mean,        # length q
    :predictions,   # length n*q (row-major)
  )

  # Fit a SIMPLS PLS regression on row-major matrices x and y.
  #
  # x must be an Array of Float of length n*p; y must have length n*q.
  # Returns a FitResult. Raises ArgumentError on shape mismatch and
  # RuntimeError on a non-zero status from libp4a.
  def self.pls_fit(x, y, n, p, q, n_components)
    raise ArgumentError, "x length #{x.length} != n*p (#{n * p})" if x.length != n * p
    raise ArgumentError, "y length #{y.length} != n*q (#{n * q})" if y.length != n * q
    raise ArgumentError, "n_components must be >= 1" if n_components < 1

    x_buf = x.pack("E*")                            # IEEE 754 little-endian doubles
    y_buf = y.pack("E*")
    coefs = ([0.0] * (p * q)).pack("E*")
    xm = ([0.0] * p).pack("E*")
    ym = ([0.0] * q).pack("E*")
    preds = ([0.0] * (n * q)).pack("E*")

    status = Lib.p4a_pls_fit_simple(
      x_buf, y_buf, n, p, q, n_components,
      coefs, xm, ym, preds
    )
    raise "p4a_pls_fit_simple failed with status #{status}" if status != 0

    FitResult.new(
      n, p, q, n_components,
      coefs.unpack("E*"),
      xm.unpack("E*"),
      ym.unpack("E*"),
      preds.unpack("E*")
    )
  end
end

File: bindings/lua/pls4all.lua
-- SPDX-License-Identifier: CECILL-2.1
--
-- Lua / LuaJIT binding around libp4a's `p4a_pls_fit_simple` C ABI
-- helper. Uses LuaJIT's built-in FFI library — no external rocks
-- required. Parity-gated against the shared cross-binding fixture
-- `bindings/js/test/parity_fixture.json` at machine epsilon.
--
-- Requires LuaJIT (the FFI module is LuaJIT-specific, not standard
-- Lua 5.x). The plain-Lua port would use `lua-ffi-libffi` or the
-- `alien` rock; we ship the canonical zero-dep LuaJIT path here.

local ffi = require("ffi")

ffi.cdef[[
typedef int p4a_status_t;
const char* p4a_get_version_string(void);
unsigned int p4a_get_abi_version_major(void);
unsigned int p4a_get_abi_version_minor(void);
unsigned int p4a_get_abi_version_patch(void);
p4a_status_t p4a_pls_fit_simple(
    const double* x, const double* y,
    int n, int p, int q, int n_components,
    double* coefficients_out,
    double* x_mean_out,
    double* y_mean_out,
    double* predictions_out);
]]

local function load_libp4a()
  local override = os.getenv("PLS4ALL_LIB")
  local candidates = {}
  if override and override ~= "" then table.insert(candidates, override) end
  table.insert(candidates, "p4a")          -- standard loader
  table.insert(candidates, "libp4a.so")
  table.insert(candidates, "libp4a.dylib")

  -- Walk up from this file to find a CMake dev-release build.
  local source = debug.getinfo(1, "S").source:sub(2)
  local cur = source:match("^(.*)/[^/]+$") or "."
  while cur and cur ~= "/" and cur ~= "" do
    for _, name in ipairs({ "libp4a.so", "libp4a.dylib" }) do
      local path = cur .. "/build/dev-release/cpp/src/" .. name
      table.insert(candidates, path)
    end
    local parent = cur:match("^(.*)/[^/]+$")
    if not parent or parent == cur then break end
    cur = parent
  end

  for _, path in ipairs(candidates) do
    local ok, lib = pcall(ffi.load, path)
    if ok then return lib end
  end
  error("could not load libp4a (set PLS4ALL_LIB to its absolute path)")
end

local lib = load_libp4a()

local M = {}

--- Returns the libp4a runtime version, e.g. "0.92.0+abi.1.13.0".
function M.version()
  local raw = lib.p4a_get_version_string()
  if raw == nil then return "" end
  return ffi.string(raw)
end

--- Returns {major, minor, patch} from the libp4a ABI.
function M.abi_version()
  return {
    tonumber(lib.p4a_get_abi_version_major()),
    tonumber(lib.p4a_get_abi_version_minor()),
    tonumber(lib.p4a_get_abi_version_patch()),
  }
end

--- Fit a SIMPLS PLS regression on row-major matrices x and y.
--
-- `x` and `y` are Lua sequences (1-indexed tables) of floats with
-- length n*p and n*q respectively. Returns a table with fields
-- `coefficients`, `x_mean`, `y_mean`, `predictions`, each as a Lua
-- sequence. The C ABI is row-major; the order of `x` is
-- (sample_0_feat_0, sample_0_feat_1, ..., sample_1_feat_0, ...).
function M.pls_fit(x, y, n, p, q, n_components)
  if #x ~= n * p then
    error(string.format("x length %d != n*p (%d)", #x, n * p))
  end
  if #y ~= n * q then
    error(string.format("y length %d != n*q (%d)", #y, n * q))
  end
  if n_components < 1 then
    error("n_components must be >= 1")
  end

  local x_buf = ffi.new("double[?]", n * p, x)
  local y_buf = ffi.new("double[?]", n * q, y)
  local coefs = ffi.new("double[?]", p * q)
  local xm    = ffi.new("double[?]", p)
  local ym    = ffi.new("double[?]", q)
  local preds = ffi.new("double[?]", n * q)

  local status = lib.p4a_pls_fit_simple(
    x_buf, y_buf, n, p, q, n_components, coefs, xm, ym, preds)
  if status ~= 0 then
    error(string.format("p4a_pls_fit_simple failed with status %d", status))
  end

  local function to_table(buf, len)
    local t = {}
    for i = 0, len - 1 do t[i + 1] = buf[i] end
    return t
  end

  return {
    n = n, p = p, q = q, n_components = n_components,
    coefficients = to_table(coefs, p * q),
    x_mean       = to_table(xm,    p),
    y_mean       = to_table(ym,    q),
    predictions  = to_table(preds, n * q),
  }
end

return M
